torch/csrc/lazy/ts_backend/ts_backend_impl.cpp

```diff
@@ -0,0 +1,244 @@
+    std::vector<torch::lazy::ComputationPtr> instances) const {
+  for (const auto& instance : instances) {
+    auto ts_computation =
+        static_cast<torch::lazy::TSComputation*>(instance.get());
+  }
+  return instances;
+}
+
+std::vector<torch::lazy::BackendDataPtr> TSBackendImpl::ExecuteComputation(
+    torch::lazy::Computation& computation,
+    c10::ArrayRef<torch::lazy::BackendDataPtr> arguments,
+    const torch::lazy::BackendDevice& device) const {
+  torch::jit::GraphExecutor& graph_executor =
+      static_cast<torch::lazy::TSComputation&>(computation).graph_executor();
+  std::vector<torch::jit::IValue> stack;
+  for (const auto& argument : arguments) {
+    const auto ts_data = std::static_pointer_cast<TSData>(argument);
+    if (ts_data->scalar.has_value()) {
+      stack.emplace_back(ts_data->scalar.value());
+    } else {
+      // TODO(whc) should this check be made more general? it's written somewhat
+      // oddly
+      CHECK(
+          (c10::DeviceType)default_device_type_.type != at::kCUDA ||
+          ts_data->data().device().type() == at::kCUDA);
+      stack.emplace_back(ts_data->data());
+    }
+  }
+  graph_executor.run(stack);
+  std::vector<torch::lazy::BackendDataPtr> results;
+  for (torch::jit::IValue component : stack) {
+    at::Tensor result = component.toTensor();
+    at::IntArrayRef result_sizes = result.sizes();
+    torch::lazy::Shape shape(
+        result.scalar_type(),
+        std::vector<int64_t>(result_sizes.begin(), result_sizes.end()));
+    results.push_back(std::make_shared<TSData>(result, shape, device));
+  }
+  return results;
+}
+
+std::vector<torch::lazy::BackendDevice> TSBackendImpl::GetBackendDevices()
+    const {
+  std::vector<torch::lazy::BackendDevice> devices;
+  // TODO(whc) figure out how to query available devices from pytorch
+  devices.emplace_back(GetBackendDevice(c10::Device(c10::kCPU, 0)));
+  devices.emplace_back(GetBackendDevice(c10::Device(c10::kCUDA, 0)));
+  return devices;
+}
+
+torch::lazy::BackendDevice TSBackendImpl::GetBackendDevice(
+    c10::Device device) const {
+  // Note, we ignore the device type specified by the c10::Device since it is
+  // expected to be a virtual device (lazy::), but we need to change this when
+  // we support lazy as a mode
+  return torch::lazy::BackendDevice(GetDefaultDeviceType(), device.index());
+}
+
+void TSBackendImpl::PrepareToExit() const {}
+
+c10::DeviceType TSBackendImpl::EagerFallbackDeviceType() const {
+  // For TS backend, hardware device _is_ eager device
+  return (c10::DeviceType)GetDefaultDeviceType()->type;
+}
+
+torch::lazy::BackendImplInterface* GetTSBackendImpl() {
+  static TSBackendImpl* ts_backend_impl = new TSBackendImpl();
+  return ts_backend_impl;
+}
+
+void InitTorchScriptBackend() {
+  static std::unique_ptr<BackendRegistrar> s_registrar;
+  s_registrar = std::make_unique<BackendRegistrar>(GetTSBackendImpl());
+
+  #if !EAGER_REGISTRATION
```

Comment:
technically this should always be set to 1 for TS, right? Maybe we should just assert it.

Reply:
Yea, I think that's true.